core/src/main/java/com/linecorp/armeria/server/cors/CorsService.java

```diff
@@ -172,60 +215,45 @@ private static HttpResponse forbidden() {
         return HttpResponse.of(HttpStatus.FORBIDDEN);
     }
 
-    /**
-     * Validates if origin matches the CORS requirements.
-     *
-     * @param request the HTTP request to check
-     * @return {@code true} if origin matches
-     */
-    private boolean validateCorsOrigin(final HttpRequest request) {
-        if (config.isAnyOriginSupported()) {
-            return true;
-        }
-
-        final String origin = request.headers().get(HttpHeaderNames.ORIGIN);
-        return origin == null ||
-               NULL_ORIGIN.equals(origin) && config.isNullOriginAllowed() ||
-               config.origins().contains(Ascii.toLowerCase(origin));
-    }
-
     /**
      * Sets origin header according to the given CORS configuration and HTTP request.
      *
      * @param request the HTTP request
      * @param headers the HTTP headers to modify
      *
-     * @return {@code true} if CORS configuration matches, otherwise false
+     * @return {@code policy} if CORS configuration matches, otherwise null
      */
-    private boolean setCorsOrigin(final HttpRequest request, final HttpHeaders headers) {
+    @Nullable
+    private CorsPolicy setCorsOrigin(final HttpRequest request, final HttpHeaders headers) {
         if (!config.isEnabled()) {
-            return false;
+            return null;
         }
 
         final String origin = request.headers().get(HttpHeaderNames.ORIGIN);
         if (origin != null) {
-            if (NULL_ORIGIN.equals(origin) && config.isNullOriginAllowed()) {
+            final CorsPolicy policy = config.getPolicy(origin);
+            if (policy == null) {
+                logger.debug("There is no CORS policy configured for the request origin [{}]]", origin);
```

Comment:
Could you also log the current `ServiceRequestContext`?  You'll have to add `ServiceRequestContext` as the first parameter of `setCorsOrigin()`.